cpp/src/strings/copying/concatenate.cu

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bitmask/legacy/valid_if.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/detail/concatenate.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include "../utilities.hpp"
+
+#include <thrust/for_each.h>
+#include <thrust/transform_reduce.h>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+std::unique_ptr<column> concatenate_vertically( std::vector<strings_column_view> const& strings_columns,
+                                                rmm::mr::device_memory_resource* mr,
+                                                cudaStream_t stream )
+{
+    if( strings_columns.size() == 0 )
+        return make_empty_strings_column(mr,stream);
+
+    size_t strings_count = thrust::transform_reduce( strings_columns.begin(), strings_columns.end(),
+        [] (auto scv) { return scv.size(); }, static_cast<size_t>(0), thrust::plus<size_t>());
+    CUDF_EXPECTS( strings_count < std::numeric_limits<size_type>::max(), "total number of strings is too large for cudf column" );
+    if( strings_count == 0 )
+        return make_empty_strings_column(mr,stream);
+    size_t total_bytes = thrust::transform_reduce( strings_columns.begin(), strings_columns.end(),
+        [] (auto scv) { return scv.chars().size(); }, static_cast<size_t>(0), thrust::plus<size_t>());
+    CUDF_EXPECTS( total_bytes < std::numeric_limits<size_type>::max(), "total size of strings is too large for cudf column" );
+
+    auto execpol = rmm::exec_policy(stream);
+
+    // create chars column
+    auto chars_column = make_numeric_column( data_type{INT8}, total_bytes, mask_state::UNALLOCATED, stream, mr);
+    auto chars_view = chars_column->mutable_view();
+    auto d_new_chars = chars_view.data<char>();
+
+    // create offsets column
+    auto offsets_column = make_numeric_column( data_type{INT32}, strings_count + 1, mask_state::UNALLOCATED, stream, mr);
+    auto offsets_view = offsets_column->mutable_view();
+    auto d_new_offsets = offsets_view.data<int32_t>();
+
+    // copy over the data for all the columns
+    ++d_new_offsets;
+    size_type offset_adjust = 0;
+    for( auto column = strings_columns.begin(); column != strings_columns.end(); ++column )
```

Comment:
Fine for now, but eventually it may be necessary to optimize this to a single kernel launch rather than a bunch of sequential copies.